File: tests/topofit/matlab/test_fit.m
clear
clc


for ii=1:3,
    fname = ['../test' num2str(ii) '.dat'];
    
    fid = fopen(fname, 'r');
    raw = fread(fid, inf, 'float32');
    fclose(fid);

    nIfg = (length(raw)-4)/2;
    wvl = raw(1);
    rng = raw(2);
    inc = raw(3) * pi / 180.0;
    delz = raw(4);

    bperp = raw(5:nIfg+4);
    ph =  raw(nIfg+5:2*nIfg+4);
    cph = exp(j * ph');

    maxK = 20.0/(wvl * rng * sin(inc)/4/pi);
    n_trial = (max(bperp) - min(bperp))*maxK/(2*pi);

    [K_r, C_r, coh_r] = ps_topofit(cph, bperp, n_trial, 'n');

    out = sprintf('file=%s K=%f C=%f coh=%f K_true=%f', fname, K_r,C_r, coh_r,delz/(wvl*rng*sin(inc)/4/pi));
    disp(out)

    figure('Name', num2str(ii));
    rph = C_r + K_r * bperp;

    rph = rph - round( rph/ (2*pi)) * 2*pi;

    scatter(bperp, ph, 'r');
    hold on;
    scatter(bperp, rph,'k');
end

File: tests/topofit/test_fit.cpp
#include <iostream>
#include <armadillo>
#include <vector>
#include <string>
#include "fringe_topofit.hpp"


int main(int argc, char **argv)
{
   
    std::vector<std::string> files(4);
    files[0] = "test1.dat";
    files[1] = "test2.dat";
    files[2] = "test3.dat";

    std::complex<float> cJ(0.0, 1.0);

    for(int ii=0; ii<3; ii++)
    {
        arma::fvec alldata;
        alldata.load(files[ii], arma::raw_binary);

        int nifg = (alldata.size() - 4)/2;
        arma::fvec bperp(alldata.memptr() + 4, nifg, false);
        arma::fvec ph(alldata.memptr()+nifg+4, nifg, false);
        
        arma::cx_fvec cph(nifg);
        arma::cx_fvec resid(nifg);

        for(int kk=0; kk<nifg; kk++)
            cph[kk] = std::exp(cJ * ph[kk]);

        float wvl = alldata[0];
        float rng = alldata[1];
        float inc = alldata[2];
        float delz = alldata[3];

        float Kmod, Cph;

        topofit worker(20.0, 0.1, wvl, nifg);
        
        float coh = worker.fit(cph.memptr(), bperp.memptr(), rng, inc, resid.memptr(), Kmod, Cph);
        std::cout << "File: " << files[ii] << " K: " << Kmod
                << " C: " << Cph << " coh: " << coh 
                << " K_true: " << delz/(wvl*rng*sin(inc*M_PI/180.0)/4.0/M_PI)<< "\n";
    }
}
